Keras - Facial Recognition

### Problem Description
Implement a Deep Learning Model that recognizes faces of individual to allow access entry to a location only if individual is happy

The "Happy" rule is strictly applied, the algorithm will use pictures from the front door camera to check if the person is happy or not. The door should open only if the person is happy. 

You have gathered pictures of your friends and yourself, taken by the front-door camera. The dataset is labbeled. 

<img src="images/house-members.png" style="width:550px;height:250px;">


In [3]:
#Import Required Packages
import numpy as np
from keras import layers
from keras.layers import Input, Dense, Activation, ZeroPadding2D, BatchNormalization, Flatten, Conv2D
from keras.layers import AveragePooling2D, MaxPooling2D, Dropout, GlobalMaxPooling2D, GlobalAveragePooling2D
from keras.models import Model
from keras.preprocessing import image
from keras.utils import layer_utils
from keras.utils.data_utils import get_file
from keras.applications.imagenet_utils import preprocess_input
import pydot
from IPython.display import SVG
from keras.utils.vis_utils import model_to_dot
from keras.utils import plot_model
from kt_utils import *

import keras.backend as K
K.set_image_data_format('channels_last')
import matplotlib.pyplot as plt
from matplotlib.pyplot import imshow

%matplotlib inline

In [7]:
#Load Datasets
X_train_orig, Y_train_orig, X_test_orig, Y_test_orig, classes = load_dataset()

#Normalize image vectors, dividing by 225 the range can be described with a 0.0-1.0 where 0.0 means (0x00)
#and 1.0 means 255(0XFF)
X_train = X_train_orig/255
X_test = X_test_orig/255

# Reshape
Y_train = Y_train_orig.T
Y_test = Y_test_orig.T

print ("number of training examples = " + str(X_train.shape[0]))
print ("number of test examples = " + str(X_test.shape[0]))
print ("X_train shape: " + str(X_train.shape))
print ("Y_train shape: " + str(Y_train.shape))
print ("X_test shape: " + str(X_test.shape))
print ("Y_test shape: " + str(Y_test.shape))


number of training examples = 600
number of test examples = 150
X_train shape: (600, 64, 64, 3)
Y_train shape: (600, 1)
X_test shape: (150, 64, 64, 3)
Y_test shape: (150, 1)


### Details
* Images are of shape (64,64,3)
* Training Set: 600 Pictures
* Test Set: 150 Pictures

In [8]:
#Happy Model Built On Keras Framework

def Happy_Facial_Recognition(input_shape):
    """
    Implementation of the Happy Model
    
    Arguments:
    input_shape -- shape of the images of the dataset

    Returns:
    model -- a Model() instance in Keras
    """
    
    #Instantiate a Keras tensor
    X_input =Input(input_shape)
    
    # Zero-Padding: pads the border of X_input with zeroes
    #Padding produces an output that has the same length of the original input. Padding prevents our image from 
    #getting smaller with each convolutional layer
    X = ZeroPadding2D((3, 3))(X_input)
    
    
    # CONV -> BN -> RELU Block applied to X
    X = Conv2D(32, (7, 7), strides = (1, 1), name = 'conv0')(X)
    
    """
    Batch normalization reduces the amount by which the hidden unit values shift around.It's a form of slight
    regularization effects and adds some noise to each hidden layer's activations. To increase stability of a
    neural network, batch normalization normalizes the output of a previous activation layer by subtracting the
    batch mean and dividing by the batch standard deviation.
    """
    
    X = BatchNormalization(axis = 3, name = 'bn0')(X)
    X = Activation('relu')(X)
    
    # MAXPOOL
    """Max pooling dis utilized to down-sample the features (such as edges) that were just extracted which gives 
    a more approximate representation of where the other features are, making the network more generalizable."""
    X = MaxPooling2D((2, 2), name='max_pool')(X)
    
    
    #FLATTEN X
    X = Flatten()(X)
    X = Dense(1, activation='sigmoid', name='fc')(X)
    
    #Generate model instance
    model = Model(inputs = )
    
    return model